pulsar/impl_partition_consumer.go

```diff
@@ -238,20 +238,55 @@ func (pc *partitionConsumer) internalUnsubscribe(unsub *handleUnsubscribe) {
 	unsub.waitGroup.Done()
 }
 
+func (pc *partitionConsumer) trackMessage(msgID MessageID) error {
+	id := &pb.MessageIdData{}
+	err := proto.Unmarshal(msgID.Serialize(), id)
+	if err != nil {
+		pc.log.WithError(err).Errorf("unserialize message id error:%s", err.Error())
+		return err
+	}
+	if pc.unAckTracker != nil {
+		pc.unAckTracker.Add(id)
+	}
+	return nil
+}
+
+func (pc *partitionConsumer) increaseAvailablePermits(receivedSinceFlow uint32) error {
+	highwater := uint32(math.Max(float64(cap(pc.options.MessageChannel)/2), 1))
+	if receivedSinceFlow >= highwater {
+		if err := pc.internalFlow(receivedSinceFlow); err != nil {
+			pc.log.Errorf("Send Flow cmd error:%s", err.Error())
+			return err
+		}
+		receivedSinceFlow = 0
+	}
+	return nil
+}
+
 func (pc *partitionConsumer) Receive(ctx context.Context) (Message, error) {
+	var receivedSinceFlow uint32
+	if err := pc.internalFlow(1); err != nil {
+		pc.log.Errorf("Send Flow cmd error:%s", err.Error())
+		return nil, err
+	}
+
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	case cm, ok := <-pc.subQueue:
 		if ok {
-			id := &pb.MessageIdData{}
-			err := proto.Unmarshal(cm.ID().Serialize(), id)
+
+			// track messages by message ID
+			err := pc.trackMessage(cm.ID())
 			if err != nil {
-				pc.log.WithError(err).Errorf("unserialize message id error:%s", err.Error())
 				return nil, err
 			}
-			if pc.unAckTracker != nil {
-				pc.unAckTracker.Add(id)
+
+			// increase available permits
+			receivedSinceFlow++
+			err = pc.increaseAvailablePermits(receivedSinceFlow)
```

Comment:
please follow java implementatiaon, and put `trackMessage` and `increaseAvailablePermits` into one method called "messageProcessed",